R/tax_range_time.R

```diff
@@ -0,0 +1,178 @@
+#' unique taxa (`taxa`), taxa ID (`taxa_id`), first appearance of taxa
+#' (`FAD_ma`), last appearance of taxa (`LAD_ma`), and temporal range duration
+#' (`range_myr`) is returned.
+#'
+#' @details The temporal range(s) of taxa are calculated by extracting all
+#' unique taxa (`name` column) from the input `occdf`, and checking their first
+#' and last appearance. The temporal duration of each taxa is also calculated.
+#'
+#' @section Developer(s):
+#' Lewis A. Jones
+#' @section Reviewer(s):
+#' To be reviewed
+#' @importFrom graphics points
+#' @examples
+#' # Grab internal data
+#' occdf <- tetrapods
+#' # Temporal range
+#' tax_range_time(occdf = occdf, name = "accepted_name", plot = TRUE)
+#'
+#' @export
+tax_range_time <- function(occdf,
+                           name = "name",
+                           min_ma = "min_ma",
+                           max_ma = "max_ma",
+                           scale = "GTS2020",
+                           plot = FALSE,
+                           return_error = FALSE) {
+
+  #=== Handling errors ===
+  if (is.data.frame(occdf) == FALSE) {
+    stop("`occdf` should be a dataframe")
+  }
+
+  if (is.logical(plot) == FALSE) {
+    stop("`plot` should be logical (TRUE/FALSE)")
+  }
+
+  if (class(occdf[, max_ma]) != class(occdf[, min_ma])) {
+    stop("The class of max_ma and min_ma must be the same")
+  }
+
+  if (any(c(name, min_ma, max_ma) %in% colnames(occdf) == FALSE)) {
+    stop("Either `name`, `min_ma`, or `max_ma`, do not exist in `occdf`")
+  }
+
+  if (any(is.na(occdf[, name]))) {
+    stop("The `name` column contains NA values")
+  }
+
+  if (any(is.na(occdf[, min_ma])) || any(is.na(occdf[, max_ma]))) {
+    stop("`min_ma` and/or `max_ma` columns contain NA values")
+  }
+
+  #=== Set-up ===
+  unique_taxa <- unique(occdf[, name])
+  # Order taxa
+  unique_taxa <- unique_taxa[order(unique_taxa)]
+
+  # Character string entered
+  if (is.character(occdf[, min_ma]) && is.character(occdf[, max_ma])) {
+    # Which GTS should be used?
+    if (scale == "GTS2020") {
+      df <- palaeoverse::GTS2020
+    }
+    if (scale == "GTS2012") {
+      df <- palaeoverse::GTS2012
+    }
+    df <- subset(df, rank == "stage")
+    # Which intervals are present?
+    unique_intervals <- unique(c(occdf[, max_ma], occdf[, min_ma]))
+    # Which of these intervals are present in the GTS
+    vec <- which(df$interval_name %in% unique_intervals)
+    # If not available, stop and return error
+    if (length(vec) != length(unique_intervals)) {
+      error_vec <- which(!unique_intervals %in% df$interval_name)
+      if (return_error == TRUE) {
+        return(error_vec)
+      } else {
+      stop(
+        paste(c(
+          "Check spelling of specified max_ma and min_ma intervals.
+  Available interval names are accessible via GTS2020 and GTS2012.
+  Check rows:",
+          capture.output(print(error_vec))
+        ),
+        collapse = "\n"
+      ))
+      }
+    }
```

Comment:
For names, how does this work in practice? NAs aren't allowed, but I could imagine that the PBDB data structure could be a typical one, using 'early_interval' and 'late_interval', but with only 'early_interval' occupied in the case that an occurrence is restricted to a single named interval.

Reply:
In this case, the user should supply the min_ma and max_ma as the same (i.e. copy the data across from the early interval to late interval). I agree this requires extra handling, but I think it is good that we don't automate this as it might not be the intention of the user, and it should be flagged.